File: tests/test_mpi3rma_fortran/rget_lockall.f90
      subroutine test_mpi3rma_rget_lockall( base, array_size, DIM1, win, comm )

      use utilities, only: check_fompi_status

#if INCLUDE_TYPE==1
      implicit none

      include 'fompif.h'
#endif
#if INCLUDE_TYPE==2
      use fompi

      implicit none
#endif

      integer, intent(in) :: array_size
      integer, dimension(array_size), intent(inout) :: base
      integer, intent(in) :: DIM1
      integer, intent(in) :: win
      integer, intent(in) :: comm
  
      integer :: i
      integer :: rank
      integer :: commsize
      integer :: mpierr
      integer(kind=MPI_ADDRESS_KIND) :: start

      logical :: flag
      integer(kind=MPI_ADDRESS_KIND) :: create_flavor

      integer :: status_rma(MPI_STATUS_SIZE)
      integer, dimension(:), allocatable :: request
  
      integer :: minimum
      integer(kind=MPI_ADDRESS_KIND), dimension(:), allocatable :: base_adr
 
      character(50) :: filename

      write (filename,*) "rget_lockall.f90"
  
      call MPI_Comm_rank( comm, rank, mpierr )
      call MPI_Comm_size( comm, commsize, mpierr )

      call foMPI_Win_lock( foMPI_LOCK_EXCLUSIVE, rank, 0, win, mpierr )
      call check_fompi_status(mpierr, MPI_SUCCESS, filename, 38, rank)

      if (rank .eq. 0 ) then
        base(1:array_size) = 0
      else
        base(1) = rank
      endif

      call foMPI_Win_unlock( rank, win, mpierr )
      call check_fompi_status(mpierr, MPI_SUCCESS, filename, 47, rank)

      call foMPI_Win_get_attr( win, foMPI_WIN_CREATE_FLAVOR, create_flavor, flag, mpierr )
      call check_fompi_status(mpierr, MPI_SUCCESS, filename, 50, rank)
      if (flag .neqv. .true.) then
        write (*,*) "Didn't found foMPI_WIN_CREATE_FLAVOR keyval."
        STOP
      endif

      if ( create_flavor .eq. foMPI_WIN_FLAVOR_DYNAMIC ) then
        if (rank .eq. 0) then
          allocate( base_adr(commsize), stat=mpierr )
        endif
        call foMPI_Get_address( base(1), start, mpierr )
        call MPI_Gather( start, 1, MPI_AINT, base_adr, 1, MPI_AINT, 0, comm, mpierr )
      else
        start = 0
      endif
      call MPI_Barrier( comm, mpierr ) ! guarantees that the initialization is finished

      if (rank .eq. 0) then
        allocate( request(commsize), stat=mpierr )
 
        if (commsize-1 .gt. array_size) then
          minimum = array_size
        else
          minimum = commsize-1
        endif
    
        call foMPI_Win_lock_all(0, win, mpierr )
        call check_fompi_status(mpierr, MPI_SUCCESS, filename, 77, rank)

        do i=1,minimum
          if ( create_flavor .eq. foMPI_WIN_FLAVOR_DYNAMIC ) then
            call foMPI_Rget( base(i), 1, MPI_INTEGER, i, base_adr(i+1), 1, MPI_INTEGER, win, request(i+1), mpierr )
          else
            call foMPI_Rget( base(i), 1, MPI_INTEGER, i, start, 1, MPI_INTEGER, win, request(i+1), mpierr)
          endif
          call check_fompi_status(mpierr, MPI_SUCCESS, filename, 85, rank)
        enddo
    
        call foMPI_Win_unlock_all( win, mpierr )
        call check_fompi_status(mpierr, MPI_SUCCESS, filename, 89, rank)

        do i=1,minimum
          call foMPI_Test( request(i+1), flag, status_rma, mpierr )
          call check_fompi_status(mpierr, MPI_SUCCESS, filename, 93, rank)
          if (flag .neqv. .true.) then
            write (*,*) rank, ": request ", i, &
              " was not finished after unlock"
            STOP
          endif
        enddo
  
        deallocate( request, stat=mpierr )

        if ( create_flavor .eq. foMPI_WIN_FLAVOR_DYNAMIC ) then
          deallocate( base_adr, stat=mpierr )
        endif

        do i=1,minimum
          if( base(i) .ne. i ) then
            write (*,*) "Error on rank ", rank, " in ", trim(filename), &
              " line ", 109, ": element ", i, " is ", base(i), &
              " instead of ", i, "."
            STOP
          endif
        enddo
      endif

      call MPI_Barrier( comm, mpierr ) ! wait for the end of the accesses

      if ( rank .eq. 0 ) then
        write (*,*) "No Error"
      endif

      end subroutine test_mpi3rma_rget_lockall

File: tests/test_foMPI-NA/rget_lockall.c
#include <assert.h>

#include "fompi.h"
#include "test_mpi3rma.h"

void test_mpi3rma_rget_lockall( int* base, int DIM, foMPI_Win win, MPI_Comm comm ) {

  int i;
  int rank;
  int commsize;
  int status;
  MPI_Aint start;

  int flag, create_flavor;

  MPI_Status status_rma;
  foMPI_Request* request;
  int min;
  MPI_Aint* base_adr = NULL; /* to prevent compiler warnings */
  
  MPI_Comm_rank( comm, &rank );
  MPI_Comm_size( comm, &commsize );

  status = foMPI_Win_lock( foMPI_LOCK_EXCLUSIVE, rank, 0, win );
  _check_fompi_status(status, MPI_SUCCESS, (char*) __FILE__, __LINE__);

  base[0] = rank;

  status = foMPI_Win_unlock( rank, win );
  _check_fompi_status(status, MPI_SUCCESS, (char*) __FILE__, __LINE__);

  status = foMPI_Win_get_attr( win, foMPI_WIN_CREATE_FLAVOR, &create_flavor, &flag );
  _check_fompi_status(status, MPI_SUCCESS, (char*) __FILE__, __LINE__);
  assert( flag != 0 );

  if ( create_flavor == foMPI_WIN_FLAVOR_DYNAMIC ) {
    if (rank == 0) {
      base_adr = malloc( commsize * sizeof(MPI_Aint) );
    }
    MPI_Get_address( &base[0], &start);
    MPI_Gather( &start, 1, MPI_AINT, base_adr, 1, MPI_AINT, 0, comm );
  } else {
    start = 0;
  }
  MPI_Barrier( comm ); /* guarantees that the initialization is finished */

  if (rank == 0) {
    request = malloc( commsize * sizeof(foMPI_Request) );
 
    if (commsize > 3*DIM) {
      min = 3*DIM;
    } else {
      min = commsize;
    }
    
    status = foMPI_Win_lock_all(0, win );
    _check_fompi_status(status, MPI_SUCCESS, (char*) __FILE__, __LINE__);

    for( i=0 ; i<min ; i++ ) {
      if ( create_flavor == foMPI_WIN_FLAVOR_DYNAMIC ) {
        status = foMPI_Rget( &base[i], 1, MPI_INT, i, base_adr[i], 1, MPI_INT, win, &request[i]);
      } else {
        status = foMPI_Rget( &base[i], 1, MPI_INT, i, start, 1, MPI_INT, win, &request[i]);
      }
      _check_fompi_status(status, MPI_SUCCESS, (char*) __FILE__, __LINE__);
    }
    
    status = foMPI_Win_unlock_all( win );
    _check_fompi_status(status, MPI_SUCCESS, (char*) __FILE__, __LINE__);

    for( i=0 ; i<min ; i++ ) {
      flag = 2;
      status = foMPI_Test( &request[i], &flag, &status_rma );
      _check_fompi_status(status, MPI_SUCCESS, (char*) __FILE__, __LINE__);
      assert( (flag == 0) || (flag == 1) );
      if (flag != 1) {
        printf("%i: request %i was not finished after unlock\n", rank, i);
        exit(EXIT_FAILURE);
      }
    }
  
    free( request );

    if ( create_flavor == foMPI_WIN_FLAVOR_DYNAMIC ) {
      free ( base_adr );
    }

    for( i=0 ; i<min ; i++) {
      if( base[i] != i ) {
        fprintf(stderr, "Error on rank %i in %s line %i: element %i is %i instead of %i.\n", rank, __FILE__, __LINE__, i, base[i], i);
        exit(EXIT_FAILURE);
      }
    }
  }

  MPI_Barrier( comm ); /* wait for the end of the accesses */

  if ( rank == 0 ) {
    printf("No Error\n");
  }

  return;
}
